src/hiring.rs

```diff
@@ -0,0 +1,285 @@
+}
+
+/// OpeningStage must be default constructible because it indirectly is a value in a storage map.
+/// ***SHOULD NEVER ACTUALLY GET CALLED, IS REQUIRED TO DUE BAD STORAGE MODEL IN SUBSTRATE***
+impl<BlockNumber> Default for ApplicationStage<BlockNumber> {
+    fn default() -> Self {
+        ApplicationStage::Active
+    }
+}
+
+// An application for an actor to occupy an opening.
+#[derive(Encode, Decode, Default, Debug, Eq, PartialEq, Clone, PartialOrd, Ord)]
+pub struct Application<OpeningId, BlockNumber, StakeId> {
+    // Identifier for opening for which this application is for.
+    pub opening_id: OpeningId,
+
+    // Index of arrival across all applications for given opening,
+    // which is needed for strictly ordering applications.
+    // Starts at 0.
+    pub application_index_in_opening: u32,
+
+    // Block at which this application was added.
+    pub add_to_opening_in_block: BlockNumber,
+
+    // NB: The given staking idnetifiers have a bloated purpose,
+    // and are mutable, fix this.
+    // https://github.com/Joystream/substrate-hiring-module/issues/11
+
+    // Identifier for stake that may possibly be established for role.
+    // Will be set iff the role staking policy of the corresponding opening
+    // states so AND application is not inactive.
+    pub active_role_staking_id: Option<StakeId>,
+
+    // Identifier for stake that may possibly be established for application
+    // Will be set iff the application staking policy of the corresponding opening
+    // states so.
+    pub active_application_staking_id: Option<StakeId>,
+
+    // Status of this application
+    pub stage: ApplicationStage<BlockNumber>,
+
+    // ...
+    pub human_readable_text: Vec<u8>,
+}
+
+// How to limit the number of eligible applicants
+#[derive(Encode, Decode, Debug, Eq, PartialEq, Clone)]
+pub struct ApplicationRationingPolicy {
+    // The maximum number of applications that can be on the list at any time.
+    pub max_active_applicants: u32,
+    // How applicants will be ranked, in order to respect the maximum simultaneous application limit
+    //pub applicant_ranking: ApplicationRankingPolicy
+}
+
+#[derive(Encode, Decode, Debug, Eq, PartialEq, Clone)]
+pub enum OpeningDeactivationCause {
+    CancelledBeforeActivation,
+    CancelledAcceptingApplications,
+    CancelledInReviewPeriod,
+    ReviewPeriodExpired,
+    Filled,
+}
+
+#[derive(Encode, Decode, Debug, Eq, PartialEq, Clone)]
+pub enum ActiveOpeningStage<BlockNumber> {
+    AcceptingApplications {
+        //
+        started_accepting_applicants_at_block: BlockNumber,
+    },
+
+    //
+    ReviewPeriod {
+        started_accepting_applicants_at_block: BlockNumber,
+
+        started_review_period_at_block: BlockNumber,
+    },
+
+    //
+    Deactivated {
+        cause: OpeningDeactivationCause,
+
+        deactivated_at_block: BlockNumber,
+
+        started_accepting_applicants_at_block: BlockNumber,
+
+        started_review_period_at_block: BlockNumber,
```

Comment:
When I saw this comment, it made sense, but I could not understand how this error would be possible without a build error, because there would be no valid value to provide for this variable, as you say.

The only place this can occur is `cancel_opening`, and after staring at it forever it occurred to me that the reason there was no build error was because of a much more serious bug: _the state of the opening is not even updated!_.

Will prepare fix, please double-check my logic.

Reply:
Fixed in https://github.com/Joystream/substrate-hiring-module/pull/1/commits/3eb105705dd2f317af291e361095a0d7a3822bad

Reply:
The fact that we both missed this one, *eeek*, this module is a bit too rough, both in complexity and implementation.